File: src/codes/exposed-complexity/php/rfc_iterator.php
<?php

interface Step {
  public function execute(array &$context): void;
}

class AuthStep implements Step {
  public function execute(array &$context): void {
    echo "ユーザー認証: {$context['userId']}\n";
    $context['authenticated'] = true;
  }
}

class LoadProfileStep implements Step {
  public function execute(array &$context): void {
    if (!($context['authenticated'] ?? false)) {
      throw new Exception("未認証");
    }
    echo "プロフィール取得: {$context['userId']}\n";
    $context['profile'] = ['name' => 'Hiroshi', 'id' => $context['userId']];
  }
}

class AuditStep implements Step {
  public function execute(array &$context): void {
    echo "[監査ログ] プロフィール表示\n";
  }
}

class UserProfileFlow {
  private array $steps;

  public function __construct() {
    $this->steps = [new AuthStep(), new LoadProfileStep(), new AuditStep()];
  }

  public function execute(string $userId): void {
    $context = ['userId' => $userId];
    foreach ($this->steps as $step) {
      $step->execute($context);
    }
    print_r($context['profile']);
  }
}

// 利用例
$flow = new UserProfileFlow();
$flow->execute("user-123");

File: src/codes/exposed-complexity/py/rfc_iterator.py
from abc import ABC, abstractmethod

class Step(ABC):
    @abstractmethod
    def execute(self, context: dict):
        pass

class AuthStep(Step):
    def execute(self, context: dict):
        print(f"ユーザー認証: {context['user_id']}")
        context['authenticated'] = True

class LoadProfileStep(Step):
    def execute(self, context: dict):
        if not context.get('authenticated'):
            raise Exception("未認証")
        print(f"プロフィール取得: {context['user_id']}")
        context['profile'] = {"name": "Hiroshi", "id": context['user_id']}

class AuditStep(Step):
    def execute(self, context: dict):
        print("[監査ログ] プロフィール表示")

class UserProfileFlow:
    def __init__(self):
        self.steps: list[Step] = [AuthStep(), LoadProfileStep(), AuditStep()]

    def execute(self, user_id: str):
        context = {"user_id": user_id}
        for step in self.steps:
            step.execute(context)
        print(context["profile"])

# 利用例
flow = UserProfileFlow()
flow.execute("user-123")
